File: 204.count-primes/solution.cpp
class Solution {
public:
    int countPrimes(int n) {
        if (n <= 2) return 0;
        set<int> cache;
        for (int i = 3; i < n; i += 2) {
            bool isPrime = true;
            for (int j : cache) {
                if (j * j > i) break;
                if (i % j == 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) cache.insert(i);
        }
        return cache.size() + 1;
    }
};

File: 204.count-primes/solution.go
func countPrimes(n int) int {
	if n <= 2 {
		return 0
	} else if n == 3 {
		return 1
	}
	p := make([]bool, n)
	p[0] = true
	p[1] = true
	realP := make([]int, 0)
	for i := 2; i < n; i++ {
		if !p[i] {
			realP = append(realP, i)
		}
		for j := 0; j < len(realP); j++ {
			val := realP[j] * i
			if val >= n {
				break
			}
			p[val] = true
		}
	}
	return len(realP)
}
